### DISCUSSION TITLE
Debugging C++ inside Python (pybind11)

### DISCUSSION THREAD

**Discussion by k0nze** [mfussenegger/nvim-dap#1010] [Q&A]
Hello, 

I currently work on a project which uses a library written in C++ which can be used in Python with pybind11. To debug the C++ code, I currently use `ipython` together with `lldb`. Which works as follows:

* Start `ipython` and get the process id with `!ps aux | grep -i ipython`
* Start `lldb` in another terminal and connect to `ipyhton` with `attach --pid $PID` and `continue`
* Set a breakpoint in `lldb` to a class method, e.g. `breakpoint -n Cat::purr`
* In `ipython` launch the Python script with `run examples/example.py`
* Now `lldb` breaks at the breakpoint and I can step through the code in `lldb` which is possible but also a bit tedious.

**Is there a way to set a breakpoint through `nvim-dap` in C++ and then launch the Python script through `nvim-dap-python` and have it recognize the C++ breakpoint when running the Python script?**

PS:
I prepared a little example project [https://github.com/k0nze/nvim_cpp_python_debugging](https://github.com/k0nze/nvim_cpp_python_debugging) which can be set up as follows:

```
cd $REPO_DIR
python -m venv .venv
source .venv/bin/activate
python -m pip install .
```

An example which instantiates the C++ classes in Python can be run with:
```
python examples/example.py
```

The C++ file in which I would like to set a breakpoint in is located under:
```
src/backend/cat.cc
```

Additionally, I added a minimal `nvim` config which installs plugins into a temporary directory which can be launched with:
```
nvim -lu minimal.lua
```

If the plugins don't install automatically, you can use `:PackerSync`

When started you can launch `nvim-dap` with `nvim-dap-ui` with [SPACE]+[d] and use the usual commands for setting breakpoints, and stepping though code.

Thank you in advance!

**Comment by theHamsta** [comment]
It's not possible with nvim-dap-python (as far as I Know). But you can set the break points in your c++ files and use `lldb-vscode` or one of the other debug adapters for c++. Just use `python3` + args to start the script as the executable to debug. The c++ breakpoints get resolved as soon as the c++ module gets loaded

**Comment by k0nze** [comment]
Thank you for the tip, I'm able to pass the path of `python3` interpreter to the prompt of `nvim-dap-ui`to `lldb-vscode` which opens Python interactive mode. However, I can't figure out how to pass an argument to the python interpreter to run my Python script?

**Reply by theHamsta** [reply]
I'm using a configuration similar to https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-lldb-vscode.

I've created myself a command to launch `lldb` like I would do with gdb/lldb from command line https://github.com/theHamsta/dotfiles/blob/eca0f8ac905074b58f059bcdccf0c5cbcab3de3c/.config/nvim/lua/my_debug.lua#L143-L194 / https://github.com/theHamsta/dotfiles/blob/eca0f8ac905074b58f059bcdccf0c5cbcab3de3c/.config/nvim/lua/init.lua#L683-L708 https://github.com/theHamsta/dotfiles/blob/eca0f8ac905074b58f059bcdccf0c5cbcab3de3c/.config/nvim/lua/init.lua#L1108-L1110

Now, I can do `:DebugLLDB cmd arg1 arg2 arg3`, e.g. `:DebugLLDB python3 %` to debug the current Python file. Sorry, for my config being so messy. nvim-dap didn't have many of the convenience functions it now has back then.

**Reply by theHamsta** [reply]
Would you find this useful enough for to wrap it up in a nice form for a nvim-dap-native extension that wraps all that recipes for native debuggers with maybe some pre-debug build triggers?

**Reply by k0nze** [reply]
That would be incredibly useful for me!

**Reply by Sahas-Ananth** [reply]
Hello, I am new(ish) to Neovim, and for one of my project I have to do this. Can you help me through this? My config is a fork of `kickstart.nvim` and I use `mason-nvim-dap` which I believe does the dap configurations automatically. Thanks in advance!
